src/libraries/System.Private.CoreLib/src/System/SearchValues/Strings/Helpers/AhoCorasick.cs

```diff
@@ -0,0 +1,380 @@
+        {
+            _nodes = nodes;
+            _startingCharsAsciiBitmap = startingAsciiBitmap;
+            _maxValueLength = maxValueLength;
+        }
+
+        public readonly bool ShouldUseAsciiFastScan
+        {
+            get
+            {
+                Vector256<byte> bitmap = _startingCharsAsciiBitmap;
+
+                if (IndexOfAnyAsciiSearcher.IsVectorizationSupported && bitmap != default)
+                {
+                    // If there are a lot of starting characters such that we often find one early,
+                    // the ASCII fast scan may end up performing worse than checking one character at a time.
+                    // Avoid using this optimization if the combined frequency of starting chars is too high.
+
+                    // Combined frequency of
+                    // - All digits is ~ 5
+                    // - All lowercase letters is ~ 57.2
+                    // - All uppercase letters is ~ 7.4
+                    const float MaxCombinedFrequency = 50f;
+
+                    float frequency = 0;
+
+                    for (int i = 0; i < 128; i++)
+                    {
+                        if (IndexOfAnyAsciiSearcher.BitmapContains(ref bitmap, (char)i))
+                        {
+                            frequency += CharacterFrequencyHelper.AsciiFrequency[i];
+                        }
+                    }
+
+                    return frequency < MaxCombinedFrequency;
+                }
+
+                return false;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public readonly int IndexOfAny<TCaseSensitivity, TFastScanVariant>(ReadOnlySpan<char> span)
+            where TCaseSensitivity : struct, StringSearchValuesHelper.ICaseSensitivity
+            where TFastScanVariant : struct, IFastScan
+        {
+            if (typeof(TCaseSensitivity) == typeof(StringSearchValuesHelper.CaseInsensitiveUnicode))
+            {
+                return GlobalizationMode.UseNls
+                    ? IndexOfAnyCaseInsensitiveUnicodeNls<TFastScanVariant>(span)
+                    : IndexOfAnyCaseInsensitiveUnicodeIcuOrInvariant<TFastScanVariant>(span);
+            }
+
+            return IndexOfAnyCore<TCaseSensitivity, TFastScanVariant>(span);
+        }
+
+        private readonly int IndexOfAnyCore<TCaseSensitivity, TFastScanVariant>(ReadOnlySpan<char> span)
+            where TCaseSensitivity : struct, StringSearchValuesHelper.ICaseSensitivity
+            where TFastScanVariant : struct, IFastScan
+        {
+            if (typeof(TCaseSensitivity) == typeof(StringSearchValuesHelper.CaseInsensitiveUnicode))
+            {
+                throw new UnreachableException();
+            }
+
+            ref AhoCorasickNode nodes = ref MemoryMarshal.GetArrayDataReference(_nodes);
+            int nodeIndex = 0;
+            int result = -1;
+            int i = 0;
+
+        FastScan:
```

Comment:
This section is basically the "we're not currently in anything that could possibly be a match so search ahead as fast as possible to the first possible starting point"?  Comments would be helpful :)

Reply:
Yep, it's the "we're looking for the first possible starting character" fast-path.

More comments will follow :)